docs/proposals/workflow.md

```diff
@@ -0,0 +1,255 @@
+
+
+## Implementation
+
+This proposal introduces a new REST resource `Workflow`. A `Workflow` is represented as a
+[graph](https://en.wikipedia.org/wiki/Graph_(mathematics)), more specifically as a DAG.
+Vertices of the graph represent steps of the workflow. The workflow steps are represented via a
+`WorkflowStep` resource.
+The edges of the graph are not represented explicitly - rather they are stored as a list of
+predecessors in each `WorkflowStep` (i.e. each node).
+
+
+### Workflow
+
+A new resource will be introduced in the API. A `Workflow` is a graph.
+In the simplest case it's a graph of `Job`s but it can also
+be a graph of other entities (for example cross-cluster objects or other `Workflow`s).
+
+
+```go
+// Workflow is a directed acyclic graph
+type Workflow struct {
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata.
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the expected behavior of a Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Spec WorkflowSpec `json:"spec,omitempty"`
+
+    // Status represents the current status of the Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Status WorkflowStatus `json:"status,omitempty"`
+}
+```
+
+
+#### `WorkflowSpec`
+
+```go
+// WorkflowSpec contains Workflow specification
+type WorkflowSpec struct {
+
+    // Optional duration in seconds the workflow needs to terminate gracefully. May be decreased in delete request.
+	// Value must be non-negative integer. The value zero indicates delete immediately.
+	// If this value is nil, the default grace period will be used instead.
+	// Set this value longer than the expected cleanup time for your workflow.
+    // If downstream resources (job, pod, etc.) define their TerminationGracePeriodSeconds
+    // the biggest is taken.
+	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty"`
+
+	// Steps contains the vertices of the workflow graph.
+	Steps []WorkflowStep `json:"steps,omitempty"`
+}
+```
+
+* `spec.steps`: is an array of `WorkflowStep`s.
+* `spec.terminationGracePeriodSeconds`: is the terminationGracePeriodSeconds.
+
+### `WorkflowStep`<sup>1</sup>
+
+The `WorkflowStep` resource acts as a [union](https://en.wikipedia.org/wiki/Tagged_union) of `JobSpec` and `ObjectReference`.
+
+```go
+// WorkflowStep contains necessary information to identifiy the node of the workflow graph
+type WorkflowStep struct {
+    // Id is the identifier of the current step
+    Id string  `json:"id,omitempty"`
+
+    // Spec contains the job specificaton that should be run in this Workflow.
+	// Only one between External and Spec can be set.
+	Spec JobSpec `json:"jobSpec,omitempty"`
```

Comment:
If we do hold generic objects, then there will also need to be an object-agnostic way to determine if the created object is 'completed'.